src/mul/kernel/generic.jl

```diff
@@ -18,6 +19,14 @@ indices: array of indices from comspace
     return
 end
 
+# real U, single real state, specialize on the X gate
+@inline function subspace_mul_kernel_generic!(S::AbstractVector{T}, y, indices, U::Val{:X_test}, k::Int, offset::Int) where T
+    idx_1 = k + indices[1] + offset
+    idx_2 = k + indices[2] + offset
+    S[idx_1], S[idx_2] = S[idx_2], S[idx_1]
```

Comment:
X gate is an example of permutation gates. Maybe we need a more generic way to implement it.

Reply:
I agree with you. In metaprogramming, permutation gates should be considered as an important group, so that code can be re-used for different permutation gates in `codegen`. While in generated code, we should go as specialized as possible to maximize performance.

Reply:
@ChenZhao44  we would handle this in codegen since X needs to be specialized on its entries to remove the extra fma isntruction